### PR TITLE
Proof-of-concept Rollup integration.

### PR DESCRIPTION
**Pull request by noahlange**
Rollup integration, permitting incremental conversion of JavaScript files in `src` into TypeScript.

- removes webpack dependency
- permits consolidation of js/d.ts files
- switch to babel compiler

## What does this do?

Tweego doesn't recognize `.ts` files, and from my understanding, that's prevented all the JavaScript in `src` from being converted to TypeScript. Each of those "modules" are currently split into separate JS and TS declaration files. Likewise, Tweego doesn't recognize ES modules because everything's in the global scope.

Rollup is kind of like Webpack in that it allows you to consolidate a dependency tree into a single JS file. In my experience, Rollup is better suited for working with source code, and Webpack is better for working with TTFs and JPGs and everything else under the sun.

## How was this tested? Did you test the changes in the compiled `.html` file?

Mostly? I did a spot check and wasn't able to find any problems I couldn't reproduce after building straight from master.

## Is there a [GitHub Issue](https://github.com/ryceg/Eigengrau-s-Essential-Establishment-Generator/issues?q=is%3Aopen+is%3Aissue) that this is resolving?
Not that I am aware of.

### CODE REVIEW

**Comment by Glinkis** [comment]
Thanks for the PR!
I'll take a deeper look at this before we merge it @ryceg.

**Comment by noahlange** [comment]
> Thanks for the PR!
> I'll take a deeper look at this before we merge it @ryceg.

You bet–let me know if you have any questions or would like me to go through anything in more detail.

**Comment by ryceg** [comment]
For the uneducated among us (_cough cough_), what benefit would this bring?

**Comment by noahlange** [comment]
> For the uneducated among us (_cough cough_), what benefit would this bring?

Practically speaking:

- It would allow all the JS files in `src` to be written in TypeScript alongside the code in `lib`. So all the benefits one normally gets from TypeScript, but throughout the entire project. 🙂
- It'd make it easier to gradually restructure the application to shift generation logic out of the presentation (i.e., Twee) code. That would open up a lot of options re: ways to cut down on the package size.
- Moving generation code to a WebWorker and out of the main UI thread would grant massive performance gains for end-users.

For kicks, I did a little renaming/refactoring of the files in `/src/NPCGeneration/Relationships` to use TS and the compiler picked up on a lot of things I wouldn't have noticed.

**Comment by ryceg** [comment]
Oh, snazzy! So it still able to access `State.variables`? That's very cool indeed.

**Comment by noahlange** [comment]
> Oh, snazzy! So it still able to access `State.variables`? That's very cool indeed.

Yep! In development, the only differences between this and how it was before are a) everything in `src` can now be written in TypeScript (so long as it's imported into `main.ts`) and b) init.js is now a generated file instead of a manually-written one.

In production, and everywhere else it is—or at least should be—exactly the same.

**Review by Glinkis — COMMENTED** [review]
Thank you for the pull request! This seems very interesting, but there are a few things I've noticed.

- Type checking no longer happens in the compilation.
- The generated `init` file clutters up the `src`-folder. This is only a single file, but still.
- Module functions are not scoped properly in the output. For example:
- Is there any way to making source maps work with this? Tweego itself doesn't  handle them, which is why the `lib` folder compiled to a separate .js-file with a corresponding sourcemap.

File: src/NPCGeneration/Relationships/getFatherMother.ts [hunk 1]
```diff
@@ -1,11 +1,24 @@
...
+    getFatherMother(town: Town, npc: NPC): Parents
+  }
+}
+
+interface Parents {
+  father?: NPC
+  mother?: NPC
+}
+
+console.log('bar')
```
Comment (Glinkis) [inline src/NPCGeneration/Relationships/getFatherMother.ts:15]: Leftover debugging?

File: package.json [hunk 2]
```diff
@@ -16,10 +16,10 @@
   "homepage": "https://github.com/ryceg/Eigengrau-s-Essential-Establishment-Generator/blob/master/README.md",
   "scripts": {
     "install-compiler": "node scripts/install-compiler",
-    "build": "node scripts/build",
+    "build": "cross-env NODE_ENV=production node scripts/build",
     "start": "node scripts/build --watch",
     "test": "jest",
-    "lint": "eslint **/*.{js,ts}"
+    "lint": "eslint '{lib,src,scripts}/**/*.{js,ts}'"
```
[the remark this answers is not in the document]
Comment (noahlange) [reply package.json:22]: Unquoted globs are "expanded" by the shell before the command is invoked. Quoting the glob instructs ESLint to perform the expansion instead, which is what we want to do—running `lint` on master prints an error message, while here it'll work as intended. Perfectly reasonable to revert it to `'**/*.{js,ts}'`, however.

File: src/NPCGeneration/Relationships/getFatherMother.ts (hunk 1 above)
Comment (noahlange) [reply src/NPCGeneration/Relationships/getFatherMother.ts:15]: Yikes–you got me.

File: scripts/build.js [hunk 3]
```diff
@@ -3,98 +3,128 @@ const fs = require('fs')
...
-// Extract extra arguments.
-const [, , ...flags] = process.argv
 
-const args = [
-  '--output=./gh-pages/index.html',
-  './src',
-  '--head=./main.ejs',
-  ...flags
-]
+function tweego (args) {
```
[the remark this answers is not in the document]
Comment (noahlange) [reply scripts/build.js:43]: Noted. 👍🏻

**Comment by noahlange** [comment]
> Type checking no longer happens in the compilation.

That is correct, and it's a side-effect of moving to Babel. With a `tsconfig.json` file (which I'd highly recommend adding), you can set `"noEmit"` to true and run TSC to perform type-checking as a separate command—I've typically found this easier to work with, since most of the type-checking issues I run into are indicated by VSCode anyway.

While Babel vs. TS is partially personal preference, there are some solid reasons to go with Babel. Most importantly, it offloads the cognitive burden of "is it going to work here?" to the build tooling instead of the programmer. With [babel-preset-env](https://babeljs.io/docs/en/babel-preset-env), you can indicate browser targets (or usage statistics) and have the output generated accordingly.

I _think_ the TS compiler generates ES5 by default, which is unnecessary for most use-cases nowadays.

It also has the option to output runtime shims for some more modern features (Object.values, Object.entries, etc)—which to my understanding, TS does not.

> The generated init file clutters up the src-folder. This is only a single file, but still. At the very least it should be excluded from git, linting, and other tooling.

Makes sense to me. 👍🏻 

> Is there any way to making source maps work with this? Tweego itself doesn't handle them, which is why the lib folder compiles to a separate .js-file with a corresponding sourcemap. It's also a bit painful that it's all bundled into a single humongous html-file.

...I think the babel plugin has a `sourceMaps` option. Let me check.

I agree about the bundling issues, which is why converting this code to TS would be such a boon. The more code that can be moved to `/lib`, the better, and it's much easier to do that safely when everything's in TS.

> Is function scoping handled properly in the output file? For example, if I import two files that contains functions with the same names.

Yep–Rollup will rewrite names as needed to prevent collisions.

> What's the rationalization of using rollup over webpack, as they can both compile to a single file with an IIFE? I'm not opposed to the change, just curious.

Readability. In my experience, as a project gets more complicated and as more things are exported/imported, Webpack's output becomes significantly harder to understand if you need to dig into the bundle and figure out what's going on—which, unfortunately, I've never been able to avoid entirely.

Granted, I do prefer Webpack if working with stylesheet imports, SVGs, etc. But since that isn't the case here, I don't think Webpack's cognitive overhead is worth it.

> I guess the webpack config can be deleted?

Yep–sorry about that.

I've pushed an additional commit that addresses the issues you mentioned inline. Thanks!

File: package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (Glinkis) [reply package.json]: Ah. If it fixes the issue @ryceg was having, we go with your approach.

**Comment by Glinkis** [comment]
> With a tsconfig.json file (which I'd highly recommend adding), you can set "noEmit" to true and run TSC to perform type-checking as a separate command—I've typically found this easier to work with, since most of the type-checking issues I run into are indicated by VSCode anyway.

I see this both as a pro and a con. Having type checking in the compilation step makes incremental updates easier, as it catches any type errors you can't see in the current file in the IDE.

And we already do have a tscpnfog file in the lib folder, which we can probably move up to the root in this PR.

>While Babel vs. TS is partially personal preference, there are some solid reasons to go with Babel.

Right. With this we can probably also remove a lot of the manual shimming/compatibility fluff.

> Yep–Rollup will rewrite names as needed to prevent collisions.

Perfect.

> Granted, I do prefer Webpack if working with stylesheet imports, SVGs, etc. But since that isn't the case here, I don't think Webpack's cognitive overhead is worth it.

I agree about the readability, having had to muck around in compiled files on several occasions. This is usually not a problem if you have source maps, but *can* be.

We may start importing other kind of files in the future, but until we do I think I like using rollup.

**Comment by noahlange** [comment]
> I see this both as a pro and a con. Having type checking in the compilation step makes incremental updates easier, as it catches any type errors you can't see in the current file in the IDE.

That's a good point—it's ultimately a tradeoff. For me (personally) the reduced compile time + misc. Babel features is worth the manual `yarn run types` I run whenever I want to do a full type-check.

(Missed the `tsconfig.json` file, whoops! Never mind the bit about bundle size reduction. 🙂 )

**Comment by Glinkis** [comment]
We could add the type check as a CI step. Then we at least prevent any code that hasn't been typecjevked form being merged into master.

**Comment by Glinkis** [comment]
@noahlange Not sure how to handle this with git, but I've pushed a few updates to the rollup branch in this repo. I cant seem to file a PR to your branch, so maybe you can just pull the changes from it?

**Comment by noahlange** [comment]
> @noahlange Not sure how to handle this with git, but I've pushed a few updates to the rollup branch in this repo. I cant seem to file a PR to your branch, so maybe you can just pull the changes from it?

Yep, done. Lemme know if you need anything else.

**Comment by Glinkis** [comment]
I think I'm happy with the state of it now. There are a bunch of broken type definition files in the `src` folder due to how import/export and interface merging works, so the mandatory type checking is only currently running on the `lib` folder. I think this is fine for now, and we can gradually include/fix sections under `src` until it's all covered.

**Review by Glinkis — APPROVED** [review]

**Comment by Glinkis** [comment]
@ryceg Feel free to take a last gander, and then merge.

**Pull request merged**